torch/nn/functional.py

```diff
@@ -3091,23 +3091,29 @@ def _pad_circular(input, padding):
-                                 value,                  # type: Tensor
-                                 embed_dim_to_check,     # type: int
-                                 num_heads,              # type: int
-                                 in_proj_weight,         # type: Tensor
-                                 in_proj_bias,           # type: Tensor
-                                 bias_k,                 # type: Optional[Tensor]
-                                 bias_v,                 # type: Optional[Tensor]
-                                 add_zero_attn,          # type: bool
-                                 dropout_p,              # type: float
-                                 out_proj_weight,        # type: Tensor
-                                 out_proj_bias,          # type: Tensor
-                                 training=True,          # type: bool
-                                 key_padding_mask=None,  # type: Optional[Tensor]
-                                 need_weights=True,      # type: bool
-                                 attn_mask=None          # type: Optional[Tensor]
+def multi_head_attention_forward(query,                           # type: Tensor
+                                 key,                             # type: Tensor
+                                 value,                           # type: Tensor
+                                 embed_dim_to_check,              # type: int
+                                 num_heads,                       # type: int
+                                 in_proj_weight,                  # type: Tensor
+                                 in_proj_bias,                    # type: Tensor
+                                 bias_k,                          # type: Optional[Tensor]
+                                 bias_v,                          # type: Optional[Tensor]
+                                 add_zero_attn,                   # type: bool
+                                 dropout_p,                       # type: float
+                                 out_proj_weight,                 # type: Tensor
+                                 out_proj_bias,                   # type: Tensor
+                                 training=True,                   # type: bool
+                                 key_padding_mask=None,           # type: Optional[Tensor]
+                                 need_weights=True,               # type: bool
+                                 attn_mask=None,                  # type: Optional[Tensor]
+                                 use_separate_proj_weight=False,  # type: bool
+                                 q_proj_weight=None,              # type: Optional[Tensor]
+                                 k_proj_weight=None,              # type: Optional[Tensor]
+                                 v_proj_weight=None,              # type: Optional[Tensor]
+                                 static_k=None,                   # type: Optional[Tensor]
```

Comment:
I think we had talked about maybe moving this into a separate function by splitting multi_head_attention_forward out into two, one part that does the projections and the other that consumes them. Users that want to do their own projections can then call the latter.

Reply:
But it may not really help the layout of proj_weights issues. We still have to maintain the two layouts at the same time, right?

Reply:
If you (or user) pass separate `q_proj_weight`, `k_proj_weight`, `v_proj_weight` the layout is pretty flexible, and 2 layouts are not really necessary. But this implies that (worst case) 3 gemms will be called instead of 1, which may or may not hurt performance.

Reply:
@ngimel yeap. @cpuhrsch has this performance concern so we keep the single proj_weight option here. Only use the three gemms option when key/value have different embed dimension with query.

Reply:
fwiw, we used 3 separate gemms in mlperf submission, because then .contiguous() calls here https://github.com/pytorch/pytorch/blob/6d1f0daa25a3cde30dd4f772f4157e6bdde6a720/torch/nn/functional.py#L3276-L3280 become no-ops, so it compensates for gemms potentially being less efficient. Overall which approach is preferable is size-dependent. 
Alternatively, pointers and strides can be checked for projection weights and if they point to a contiguous matrix a single gemm can be called.